File: datasets/python_source-code/Multiple-A3/organic_10.py
# Linked List Reversal - organic submission 10
# Iterator-protocol based: LinkedList implements __iter__, reversal is a
# module-level function that rebuilds links from a generator.

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class LinkedList:
    def __init__(self):
        self.head = None

    def append(self, data):
        node = Node(data)
        if self.head is None:
            self.head = node
            return
        cur = self.head
        while cur.next is not None:
            cur = cur.next
        cur.next = node

    def __iter__(self):
        cur = self.head
        while cur is not None:
            yield cur.data
            cur = cur.next

    def reverse(self):
        self.head = reverse_nodes(self.head)


def reverse_nodes(head):
    stack_of_nodes = []
    node = head
    while node is not None:
        stack_of_nodes.append(node)
        node = node.next
    new_head = None
    prev = None
    while stack_of_nodes:
        top = stack_of_nodes.pop()
        if prev is None:
            new_head = top
        else:
            prev.next = top
        prev = top
    if prev is not None:
        prev.next = None
    return new_head


if __name__ == "__main__":
    ll = LinkedList()
    for v in (1, 2, 3, 4):
        ll.append(v)
    print(list(ll))
    ll.reverse()
    print(list(ll))

File: server/datasets/controlled_benchmarks/java/multi_student_cohorts/cohort_a3/organic_10.java
// Linked List Reversal - organic submission 10
// List iteration and reversal rebuilding links.

class Node {
    int data;
    Node next;

    public Node(int data) {
        this.data = data;
        this.next = null;
    }
}

class LinkedList {
    Node head;

    public LinkedList() {
        this.head = null;
    }

    public void add(int data) {
        Node n = new Node(data);
        if (this.head == null) {
            this.head = n;
            return;
        }
        Node c = this.head;
        while (c.next != null) {
            c = c.next;
        }
        c.next = n;
    }

    public void reverse() {
        Node prev = null;
        Node cur = this.head;
        while (cur != null) {
            Node nxt = cur.next;
            cur.next = prev;
            prev = cur;
            cur = nxt;
        }
        this.head = prev;
    }

    public void display() {
        Node c = this.head;
        StringBuilder sb = new StringBuilder();
        while (c != null) {
            sb.append(c.data);
            if (c.next != null) sb.append(" -> ");
            c = c.next;
        }
        System.out.println(sb.toString());
    }

    public static void main(String[] args) {
        LinkedList ll = new LinkedList();
        for (int i = 1; i <= 5; i++) ll.add(i);
        ll.display();
        ll.reverse();
        ll.display();
    }
}
